File: DemoMethods/Basic/FacetsWithDocuments.cs
﻿using System;
using System.Linq;
using System.Web.Http;
using DemoMethods.Entities;
using DemoMethods.Indexes;
using Raven.Abstractions.Data;
using Raven.Client;

namespace DemoMethods.Basic
{
    public partial class BasicController : ApiController
    {
        [HttpGet]
        public object FacetsWithDocuments(string fromVal = "10", string toVal = "20")
        {
            try
            {
                var from = decimal.Parse(fromVal);
                var to = decimal.Parse(toVal);

                var facets = FacetRangeCreation.CreateFacets(from, to);

                using (var session = DocumentStoreHolder.Store.OpenSession())
                {
                    session.Store(new FacetSetup { Id = "facets/ProductFacet", Facets = facets });
                    session.SaveChanges();

                    var facetResults = session
                        .Query<Product, ProductsAndPriceAndSuplier>()
                        .Where(x => x.UnitsInStock > 1)
                        .ToFacets(facets);

                    return DemoUtilities.FormatRangeResults(facetResults.Results);
                }
            }
            catch (Exception e)
            {
                return e.Message;
            }
        }
    }
}

File: DemoMethods/src/main/java/net/ravendb/demo/basic/FacetsWithDocuments.java
package net.ravendb.demo.basic;

import net.ravendb.abstractions.data.FacetResults;
import net.ravendb.client.IDocumentSession;
import net.ravendb.demo.DemoUtilities;
import net.ravendb.demo.DocumentStoreHolder;
import net.ravendb.demo.MenuController;
import net.ravendb.demo.entities.Product;
import net.ravendb.demo.entities.QProduct;
import net.ravendb.demo.indexes.FacetRangeCreation;
import net.ravendb.demo.indexes.ProductsAndPriceAndSuplier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;

import java.util.List;

@Controller
public class FacetsWithDocuments {

    @RequestMapping("/Basic/FacetsWithDocuments")
    public List<FacetRangeCreation.FacetsRangesResults> facetsWithDocuments() {

        QProduct p = QProduct.product;
        try (IDocumentSession session = DocumentStoreHolder.getStore().openSession()) {
            FacetResults facetResults = session.query(Product.class, ProductsAndPriceAndSuplier.class)
                    .where(p.unitsInStock.gt(1))
                    .toFacets(MenuController.getFixedFacet());

            return DemoUtilities.formatRangeResults(facetResults.getResults());
        }
    }
}